src/Lean/Elab/Tactic/Ext.lean

```diff
@@ -0,0 +1,234 @@
+  withLocalDeclD `y (mkAppN structC params) fun y => do
+    let mut hyps := #[]
+    let fields := if flat then
+      getStructureFieldsFlattened (← getEnv) struct (includeSubobjectFields := false)
+    else
+      getStructureFields (← getEnv) struct
+    for field in fields do
+      let x_f ← mkProjection x field
+      let y_f ← mkProjection y field
+      if ← isProof x_f then
+        pure ()
+      else if ← isDefEq (← inferType x_f) (← inferType y_f) then
+        hyps := hyps.push (field, ← mkEq x_f y_f)
+      else
+        hyps := hyps.push (field, ← mkHEq x_f y_f)
+    k params x y hyps
+
+/--
+Creates the type of the extensionality lemma for the given structure,
+elaborating to `x.1 = y.1 → x.2 = y.2 → x = y`, for example.
+-/
+scoped elab "ext_type% " flat:term:max ppSpace struct:ident : term => do
+  withExtHyps (← resolveGlobalConstNoOverloadWithInfo struct) flat fun params x y hyps => do
+    let ty := hyps.foldr (init := ← mkEq x y) fun (f, h) ty =>
+      mkForall f BinderInfo.default h ty
+    mkForallFVars (params |>.push x |>.push y) ty
+
+/-- Make an `Iff` application. -/
+def mkIff (p q : Expr) : Expr := mkApp2 (mkConst ``Iff) p q
+
+/-- Make an n-ary `And` application. `mkAndN []` returns `True`. -/
+def mkAndN : List Expr → Expr
+  | [] => mkConst ``True
+  | [p] => p
+  | p :: ps => mkAnd p (mkAndN ps)
+
+/--
+Creates the type of the iff-variant of the extensionality lemma for the given structure,
+elaborating to `x = y ↔ x.1 = y.1 ∧ x.2 = y.2`, for example.
+-/
+scoped elab "ext_iff_type% " flat:term:max ppSpace struct:ident : term => do
+  withExtHyps (← resolveGlobalConstNoOverloadWithInfo struct) flat fun params x y hyps => do
+    mkForallFVars (params |>.push x |>.push y) <|
+      mkIff (← mkEq x y) <| mkAndN (hyps.map (·.2)).toList
+
+macro_rules | `(declare_ext_theorems_for $[(flat := $f)]? $struct:ident $(prio)?) => do
+  let flat := f.getD (mkIdent `true)
+  let names ← Macro.resolveGlobalName struct.getId.eraseMacroScopes
+  let name ← match names.filter (·.2.isEmpty) with
+    | [] => Macro.throwError s!"unknown constant {struct}"
+    | [(name, _)] => pure name
+    | _ => Macro.throwError s!"ambiguous name {struct}"
+  let extName := mkIdentFrom struct (canonical := true) <| name.mkStr "ext"
+  let extIffName := mkIdentFrom struct (canonical := true) <| name.mkStr "ext_iff"
+  `(@[ext $(prio)?] protected theorem $extName:ident : ext_type% $flat $struct:ident :=
+      fun {..} {..} => by intros; subst_eqs; rfl
+    protected theorem $extIffName:ident : ext_iff_type% $flat $struct:ident :=
+      fun {..} {..} =>
+        ⟨fun h => by cases h; split_ands <;> rfl,
+         fun _ => by (repeat cases ‹_ ∧ _›); subst_eqs; rfl⟩)
+
+/-- Apply a single extensionality lemma to `goal`. -/
+def applyExtLemma (goal : MVarId) : MetaM (List MVarId) := goal.withContext do
+  let tgt ← goal.getType'
+  unless tgt.isAppOfArity ``Eq 3 do
+    throwError "applyExtLemma only applies to equations, not{indentExpr tgt}"
```

Comment:
Would it be more user-friendly to also check here that the thing being applied is `Eq` or `HEq`?

Reply:
I'm not sure what you mean here.

Reply:
I'm also not sure what I mean - I think I read it incorrectly the first time around, and missed the `Eq` on line 95. Sorry for the false alarm :)